dlib/unicode/unicode.h

```diff
@@ -499,21 +605,71 @@ namespace dlib
     {
         using namespace unicode_helpers;
         ustring temp;
-        std::istringstream sin(str);
 
         temp.reserve(str.size());
 
-        int status;
+        int status = 0;
         unichar ch;
-        while ( (status = u8_to_u32(ch,sin)) > 0)
-            temp.push_back(ch);
+
+        for (size_t i = 0; i < str.size();)
+        {
+            status = u8_to_u32(ch, str.begin() + status, str.end());
+            if (status > 0)
+            {
+                temp.push_back(ch);
+                i += status;
+            }
+            else
+            {
+                break;
+            }
+        }
 
         if (status < 0)
             throw invalid_utf8_error();
 
         return temp;
     }
 
+    template <typename forward_iterator, typename unary_op>
+    inline void convert_utf8_to_utf32(forward_iterator ibegin, forward_iterator iend, unary_op op)
+    {
+        unichar ch;
+        int status = 0;
+
+        // ugly hack
+        if (sizeof(*ibegin) == sizeof(unichar))
```

Comment:
Yeah there are lots of ways to go about this.  You can use SFINAE (https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error#:~:text=Substitution%20failure%20is%20not%20an%20error%20(SFINAE)%20is%20a%20principle,to%20describe%20related%20programming%20techniques.) to control which overload of `convert_utf8_to_utf32()` is invoked.

But seems like 
```
if (!std::is_same<std::decay_t<decltype(*ibegin)>, char>::value) 
```
here would be sufficient?

Reply:
Well, probably put this at the top
```
using char_type = std::decay_t<decltype(*ibegin)>;
```

Then also do
```
static_assert (std::is_same<char_type, char>::value || std::is_same<char_type, unichar>::value);
```

Sine it's got to be one of those types or this doesn't really make sense.

Reply:
Ah, yes that worked, but having put this method inside the `font::draw_string` method, sometimes it's called with `wchar_t`, so that wouldn't compile.

Maybe I should add a conversion from wchar_t to UTF-32 in this method, too… and just name it `convert_to_utf32`?

Reply:
Ah, yeah I guess so.  I need to go look at this code more.  I forgot there were 3 char types floating around in there :confused: